src/libPMacc/include/traits/GetUniqueTypeId.hpp

```diff
@@ -57,27 +60,47 @@ const uint64_t GetUniqueTypeId<T_Type>::id = uint64_t(&GetUniqueTypeId<T_Type>::
 
 /** Get a unique id of a type
  *
- * - generate a unique 64bit id of a type at *runtime*
+ * - generate a unique id of a type at *runtime*
  * - the id of a type is equal on each instance of a process
  *
  * @tparam T_Type any object (class or typename)
- *
- * @treturn ::uid
+ * @tparam T_ResultType result type
  */
-template<typename T_Type>
+template<typename T_Type, typename T_ResultType = uint64_t>
 struct GetUniqueTypeId
 {
-    static const uint64_t uid;
-};
 
-/** instantiation of traits::GetUniqueTypeId
- *
- * - create a instance of `traits::GetUniqueTypeId` and initialize the uid
- * - map `detail::GetUniqueTypeId<T>` to a range [1; 2^64-1]
- *   `uid = unique_id_of_current_type - base_unique_id`
- */
-template<typename T_Type>
-const uint64_t GetUniqueTypeId<T_Type>::uid = detail::GetUniqueTypeId<T_Type>::id - detail::GetUniqueTypeId<uint8_t>::id;
+    /** create unique id
+     *
+     * @param maxValue largest allowed id
+     */
+    static const T_ResultType uid(uint64_t maxValue = uint64_t(1) << (sizeof (T_ResultType) * 8) - uint64_t(1))
```

Comment:
That is just the same I previously said:
A byte in C++ is !not! necessarily 8 bits width. Therefore using a hard-coded 8 is not portable.
A ```char``` is defined to the smallest machine adressable unit and is therefore equivalent to a byte. ```sizeof(char)==1```. ```CHAR_BIT``` is the size of a byte in bits and is 8 or greater. 
See [here](https://isocpp.org/wiki/faq/intrinsic-types#sizeof-char).